### PR TITLE
fix(logging): revise automatic resource detection for ingested logs

### PR DESCRIPTION
**Pull request by minherz**
Add deterministic method for detection monitored resource environment using file system.
Remove requirement for the specific call order of the resource detection methods.
Refactor code to improve readability.
Verify parent parameter to match the expected format.
Align global resource creation with the existing monitored resource types.

### CODE REVIEW

File: logging/internal/environment.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+// ResourceAtttributesGetter declares interface to lookup environment variables and metadata attributes
+type ResourceAtttributesGetter interface {
+	EnvVar(name string) (string, bool)
+	Metadata(path string) (string, bool)
+	ReadAll(path string) (string, error)
+}
+
+var getter ResourceAtttributesGetter = &defaultResourceGetter{
+	metaClient: metadata.NewClient(&http.Client{
```
Comment (losalex) [inline logging/internal/environment.go:36]: Is `http,Client` used to download a file from remote location?

File: logging/resource.go [hunk 2]
```diff
@@ -34,165 +33,172 @@ type commonResource struct{ *mrpb.MonitoredResource }
...
+var detectedResource = struct {
+	pb    *mrpb.MonitoredResource
+	attrs internal.ResourceAtttributesGetter
+	once  *sync.Once
+}{
+	attrs: internal.ResourceAttributes(),
+	once:  new(sync.Once),
+}
+
+func metadataProjectID() (string, bool) {
```
Comment (losalex) [inline logging/resource.go:45]: What happens if metadata file is corrupted? Is there a fallback for lookup in env variables?

File: logging/resource.go [hunk 3]
```diff
@@ -34,165 +33,172 @@ type commonResource struct{ *mrpb.MonitoredResource }
...
+	_, versionOK := detectedResource.attrs.EnvVar("GAE_VERSION")
+	_, instanceOK := detectedResource.attrs.EnvVar("GAE_INSTANCE")
+	return serviceOK && versionOK && instanceOK
 }
 
 func detectAppEngineResource() *mrpb.MonitoredResource {
-	projectID, err := metadata.ProjectID()
-	if err != nil {
+	projectID, ok := metadataProjectID()
+	if !ok {
```
Comment (losalex) [inline logging/resource.go:75]: Correct me if I wrong, but failed metadata reading will make code to bail out, right? If thats a case, why not to use alternate way to retrieve projectID like `os.Getenv("GOOGLE_CLOUD_PROJECT")`?

File: logging/internal/environment.go (hunk 1 above)
Comment (minherz) [reply logging/internal/environment.go:36]: relevance?

File: logging/resource.go (hunk 2 above)
Comment (minherz) [reply logging/resource.go:45]: There is no metadata file. Please, see https://cloud.google.com/compute/docs/metadata/overview

Comment (losalex) [reply logging/resource.go:45]: I believe adding this link to code would be a great reference!

File: logging/resource.go (hunk 3 above)
Comment (losalex) [reply logging/resource.go:75]: Talked offline

File: logging/internal/environment.go (hunk 1 above)
Comment (losalex) [reply logging/internal/environment.go:36]: Talked offline

**Review by losalex — APPROVED** [review]

File: logging/internal/environment.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+	"net"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	"cloud.google.com/go/compute/metadata"
+)
+
+// ResourceAtttributesGetter declares interface to lookup environment variables and metadata attributes
```
Comment (codyoss) [inline logging/internal/environment.go:28]: nit: no need to say it's is an interface the type says that. Please end godoc with a persiod.(This applies elsewhere too)

File: logging/internal/environment.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+// Metadata uses metadata package Client.Get() to lookup for metadata attributes by path
+func (g *defaultResourceGetter) Metadata(path string) (string, bool) {
+	val, err := g.metaClient.Get(path)
+	if err != nil {
+		return val, false
+	}
+	return strings.TrimSpace(val), true
+}
+
+// ReadAll read all content of the file as a string
```
Comment (codyoss) [inline logging/internal/environment.go:68]: nit: reads

File: logging/internal/environment.go [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+	"time"
+
+	"cloud.google.com/go/compute/metadata"
+)
+
+// ResourceAtttributesGetter declares interface to lookup environment variables and metadata attributes
+type ResourceAtttributesGetter interface {
+	EnvVar(name string) (string, bool)
+	Metadata(path string) (string, bool)
+	ReadAll(path string) (string, error)
```
Comment (codyoss) [inline logging/internal/environment.go:32]: Feels odd that this one returns an error and the others are bools. I wonder if you need a second return value at all. Could yoou cue off the string being "" to be a failed lookup? That should never be valid for you I think.

File: logging/resource.go [hunk 7]
```diff
@@ -34,165 +33,172 @@ type commonResource struct{ *mrpb.MonitoredResource }
...
+}{
+	attrs: internal.ResourceAttributes(),
+	once:  new(sync.Once),
+}
+
+func metadataProjectID() (string, bool) {
+	return detectedResource.attrs.Metadata("project/project-id")
+}
+
+func metadataZone() (string, bool) {
```
Comment (codyoss) [inline logging/resource.go:49]: nit: all of these functions that use detectedResource feel like they should instead be methods on that type. Having functions work on global

File: logging/resource_test.go [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+	"sync"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	mrpb "google.golang.org/genproto/googleapis/api/monitoredres"
+)
+
+const (
+	there               = "non-empty-value"
```
Comment (codyoss) [inline logging/resource_test.go:28]: nit: there seems out of place. Maybe something?

**Review by codyoss — CHANGES_REQUESTED** [review]
Just a few small things, overall looks good. Thanks!

File: logging/resource_test.go (hunk 8 above)
Comment (minherz) [reply logging/resource_test.go:28]: this is used as a non empty value. the value itself is validate but is not use. I will place `"anyvalue"` instead

**Pull request merged**